### PR TITLE
added schema support for UDFs

### PR DESCRIPTION
**Pull request by chaudum**

### CODE REVIEW

**Comment by chaudum** [comment]
creating a function without explicit schema will store the function in the `doc` schema if no `sessionContext.defaultSchema` is present.
however, the lookup of a function without explicit schema will resolve the function only in the `sessionContext.defaultSchema` (also if not present) after looking at built-ins.
this results in a weird behaviour when not setting a default schema (e.g. using `crash`):

```sql
CREATE FUNCTION fn() RETURNS DOUBLE ...;
```

`CREATE OK`

```sql
SELECT fn();
```

`SQLActionException[UnsupportedFeatureException: unknown function: fn()]`

**Comment by chaudum** [comment]
defaultSchema behaviour has been fixed with https://github.com/crate/crate/pull/5284/commits/1c72bced1179df74f04f22649c16e49781097374

File: blackbox/docs/sql/administration/user_defined_functions.txt [hunk 1]
```diff
@@ -127,11 +135,12 @@ This would overload our ``multiply`` function with more arguments::
 .. NOTE::
 
    It is considered bad practice to create functions that have the same name
-   as the CrateDB built-in functions.
-
-.. hide:
+   as the CrateDB built-in functions. If you call a function without a schema
+   name, CrateDB will look up the builtin functions first and then the
+   user-defined functions in the session schema. Therefore a built-in function
```
Comment (bogensberger) [inline blackbox/docs/sql/administration/user_defined_functions.txt]: maybe link to the session schema doc

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 2]
```diff
@@ -98,9 +101,15 @@ public FunctionImplementation getBuiltin(String name, List<DataType> argumentsTy
      * @return a function implementation.
      * @throws UnsupportedOperationException if no implementation is found.
      */
-    public FunctionImplementation getUserDefined(String name, List<DataType> argumentsTypes) throws UnsupportedOperationException {
-        FunctionResolver dynamicResolver = udfFunctionResolvers.get().get(name);
-        FunctionImplementation impl = resolveFunctionForArgumentTypes(argumentsTypes, dynamicResolver);
+    public FunctionImplementation getUserDefined(String schema, String name, List<DataType> argumentsTypes) throws UnsupportedOperationException {
+        if (schema == null) {
+            throw createUnknownFunctionException(name, argumentsTypes);
```
Comment (bogensberger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: i would move this check down to `getQualified`. `getUserDefined` shouldn't be called with a `null` schema.

File: sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java [hunk 3]
```diff
@@ -41,7 +42,7 @@
 
 import static org.hamcrest.CoreMatchers.is;
 
-@ESIntegTestCase.ClusterScope(numDataNodes = 2, numClientNodes = 0, randomDynamicTemplates = false)
+@ESIntegTestCase.ClusterScope(numDataNodes = 1, numClientNodes = 0, randomDynamicTemplates = false)
```
Comment (bogensberger) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: why is that necessary? the tests should also run in a multi node environment. We already had a bug which occured only in a multi node env.

Comment (chaudum) [inline sql/src/test/java/io/crate/integrationtests/UserDefinedFunctionsIntegrationTest.java]: good spot! did not change it deliberately

File: sql/src/main/java/io/crate/metadata/Functions.java (hunk 2 above)
Comment (chaudum) [inline sql/src/main/java/io/crate/metadata/Functions.java]: true. forgot to remove it after I changed the defaultSchema on master
